File: src/main/java/controller/RaceController.java
package controller;

import static Message.ErrorMessage.GAME_ERROR_MESSAGE;

import domain.Car;
import domain.Cars;
import domain.TryCount;
import java.io.IOException;
import java.util.List;
import utils.RandomNumberGenerator;
import view.InputView;
import view.OutputView;

public class RaceController {

    public static void main(String[] args) {
        try {
            Cars cars = registerCars();
            playGame(cars);
            showWinners(cars);
        } catch (IOException e) {
            throw new IllegalStateException(GAME_ERROR_MESSAGE.getValue());
        } catch (RuntimeException e) {
            OutputView.printMessage(e.getMessage());
        }
    }

    private static Cars registerCars() throws IOException {
        List<String> carNames = InputView.inputCarNames();
        Cars cars = new Cars();
        for (String carName : carNames) {
            cars.add(new Car(carName));
        }
        return cars;
    }

    private static void playGame(Cars cars) throws IOException {
        TryCount count = TryCount.from(InputView.inputTryCount());
        while (!count.isZero()) {
            cars.move(new RandomNumberGenerator());
            count = count.decrease();
            OutputView.printRoundResult(cars.getPositionPerCar());
        }
    }

    private static void showWinners(Cars cars) {
        Cars winners = cars.calculateWinner();
        OutputView.printWinner(winners.getNames());
    }
}

File: src/main/kotlin/controller/RaceController.kt
package controller

import domain.Car
import domain.Cars
import domain.TryCount.Companion.from
import message.ErrorMessage
import utils.RandomNumberGenerator
import view.InputView.inputCarNames
import view.InputView.inputTryCount
import view.OutputView.printMessage
import view.OutputView.printRoundResult
import view.OutputView.printWinner
import java.io.IOException

class RaceController {

    fun start() {
        try {
            val cars = registerCars()
            playGame(cars)
            showWinners(cars)
        } catch (e: IOException) {
            throw IllegalStateException(ErrorMessage.GAME_ERROR_MESSAGE.value)
        } catch (e: RuntimeException) {
            printMessage(e.message!!)
        }
    }

    @Throws(IOException::class)
    private fun registerCars(): Cars {
        val carNames = inputCarNames()
        val cars = Cars()
        for (carName in carNames) {
            cars.add(Car(carName))
        }
        return cars
    }

    @Throws(IOException::class)
    private fun playGame(cars: Cars) {
        var count = from(inputTryCount())
        while (!count.isZero) {
            cars.move(RandomNumberGenerator())
            count = count.decrease()
            printRoundResult(cars.positionPerCar)
        }
    }

    private fun showWinners(cars: Cars) {
        val winners = cars.calculateWinner()
        printWinner(winners.names)
    }
}
